commons/src/main/java/io/aiven/kafka/connect/common/source/AbstractSourceRecordIterator.java

```diff
@@ -0,0 +1,358 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.connect.common.source;
+
+import javax.validation.constraints.NotNull;
+
+import java.io.InputStream;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import org.apache.kafka.connect.data.SchemaAndValue;
+
+import io.aiven.kafka.connect.common.config.SourceCommonConfig;
+import io.aiven.kafka.connect.common.source.input.Transformer;
+import io.aiven.kafka.connect.common.source.input.utils.FilePatternUtils;
+import io.aiven.kafka.connect.common.source.task.Context;
+import io.aiven.kafka.connect.common.source.task.DistributionStrategy;
+import io.aiven.kafka.connect.common.source.task.DistributionType;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import org.apache.commons.io.function.IOSupplier;
+import org.slf4j.Logger;
+
+/**
+ * Iterator that processes cloud storage items and creates Kafka source records. Supports multiple output formats.
+ *
+ * @param <N>
+ *            the native object type.
+ * @param <K>
+ *            the key type for the native object.
+ * @param <O>
+ *            the OffsetManagerEntry for the iterator.
+ * @param <T>
+ *            The source record for the client type.
+ *
+ */
+public abstract class AbstractSourceRecordIterator<N, K extends Comparable<K>, O extends OffsetManager.OffsetManagerEntry<O>, T extends AbstractSourceRecord<N, K, O, T>>
+        implements
+            Iterator<T> {
+    /** The OffsetManager that we are using */
+    private final OffsetManager<O> offsetManager;
+
+    /** The configuration for the source */
+    private final SourceCommonConfig sourceConfig;
+    /** The transformer for the data conversions */
+    private final Transformer transformer;
+    /** the taskId of this running task */
+    private final int taskId;
+
+    /**
+     * The inner iterator to provides a base AbstractSourceRecord for a storage item that has passed the filters and
+     * potentially had data extracted.
+     */
+    private Iterator<T> inner;
+    /**
+     * The outer iterator that provides an AbstractSourceRecord for each record contained by the storage item identified
+     * by the inner record.
+     */
+    private Iterator<T> outer;
+    /** The topic(s) which have been configured with the 'topics' configuration */
+    private final Optional<String> targetTopics;
+    /** Check if the native item key is part of the 'target' files configured to be extracted from Azure */
```

Comment:
The only spot I found where you missed some abstractification!  Good job moving these to reusable templated classes :D